crates/storage/src/index_map.rs

```diff
@@ -0,0 +1,451 @@
+                index.save(storage, &pk, updated)?;
+            }
+            self.primary.save(storage, key, updated)?;
+        } else {
+            self.primary.remove(storage, key);
+        }
+        Ok(())
+    }
+}
+
+impl<'a, K, T, I, E: Encoding<T>> IndexedMap<'a, K, T, I, E>
+where
+    K: MapKey + Clone,
+    I: IndexList<T>,
+    T: Clone,
+    E: Encoding<T>,
+{
+    pub fn update<A, Err>(&'a self, storage: &mut dyn Storage, key: K, action: A) -> Result<T, Err>
+    where
+        A: FnOnce(Option<T>) -> Result<T, Err>,
+        Err: From<StdError>,
+    {
+        let input = self.may_load(storage, key.clone())?;
+        let old_val = input.clone();
+        let output = action(input)?;
+        self.replace(storage, key, Some(&output), old_val.as_ref())?;
+        Ok(output)
+    }
+}
+
+// ----------------------------------- tests -----------------------------------
+
+#[cfg(test)]
+mod tests {
+
+    use {
+        crate::{Borsh, Encoding, Index, IndexList, IndexedMap, MultiIndex, Proto, UniqueIndex},
+        borsh::{BorshDeserialize, BorshSerialize},
+        grug_types::{MockStorage, Order, StdError, StdResult},
+        prost::Message,
+        std::{str::from_utf8, vec},
+        test_case::test_case,
+    };
+
+    #[derive(BorshSerialize, BorshDeserialize, PartialEq, Eq, PartialOrd, Ord, Message, Clone)]
+    struct Foo {
+        #[prost(string, tag = "1")]
+        pub name: String,
+        #[prost(string, tag = "2")]
+        pub surname: String,
+        #[prost(uint32, tag = "3")]
+        pub id: u32,
+    }
+
+    impl Foo {
+        pub fn new(name: &str, surname: &str, id: u32) -> Self {
+            Foo {
+                name: name.to_string(),
+                surname: surname.to_string(),
+                id,
+            }
+        }
+    }
+
+    struct FooIndexes<'a, E: Encoding<Foo> = Borsh> {
+        pub name: MultiIndex<'a, String, Foo, u64, E>,
+        pub name_surname: MultiIndex<'a, (String, String), Foo, u64, E>,
+        pub id: UniqueIndex<'a, u32, Foo, u64, E>,
+    }
+
+    impl<'a, E: Encoding<Foo>> IndexList<Foo> for FooIndexes<'a, E> {
+        fn get_indexes(&self) -> Box<dyn Iterator<Item = &'_ dyn Index<Foo>> + '_> {
+            let v: Vec<&dyn Index<Foo>> = vec![&self.name, &self.name_surname, &self.id];
+            Box::new(v.into_iter())
+        }
+    }
+
+    fn foo<'a, E: Encoding<Foo>>() -> IndexedMap<'a, u64, Foo, FooIndexes<'a, E>, E> {
+        let indexes = FooIndexes::<E> {
+            name: MultiIndex::new(|_, data| data.name.clone(), "pk_namespace", "name"),
+            name_surname: MultiIndex::new(
+                |_, data| (data.name.clone(), data.surname.clone()),
+                "pk_namespace",
+                "name_surname",
+            ),
+            id: UniqueIndex::new(|data| data.id, "unique_ns"),
+        };
+
+        IndexedMap::<u64, Foo, FooIndexes<'a, E>, E>::new("pk_namespace", indexes)
+    }
```

Comment:
`IndexedMap`s can be constants, so it's unnecessary to create them using a function:

```rust
const FOO: IndexedMap<u64, Foo, FooIndexes> = IndexedMap::new("pk_namespace", FooIndexes {
    name: MultiIndex::new(|_, data| data.name.clone(), "pk_namespace", "name"),
    name_surname: MultiIndex::new(
        |_, data| (data.name.clone(), data.surname.clone()),
        "pk_namespace",
        "name_surname",
    ),
    id: UniqueIndex::new(|data| data.id, "unique_ns"),
});
```

Reply:
In this example we need `E: Encoding` to properly test with different encoding schema

Reply:
> In this example we need `E: Encoding` to properly test with different encoding schema

Understood